File: src/main/scala/com/avast/kluzo/TraceId.scala
package com.avast.kluzo

import java.util.concurrent.ThreadLocalRandom

case class TraceId(value: String) extends AnyVal

object TraceId {

  final val Length: Int = 10

  final val AllowedChars: Array[Char] = (('a' to 'z') ++ ('0' to '9')).toArray

  private val numberOfAllowedChars = AllowedChars.length

  /** This array is just used to map over it. */
  private val helperArray = new Array[Unit](numberOfAllowedChars)

  def generate: TraceId = {
    val rnd = ThreadLocalRandom.current
    def randomChar = (_: Unit) => AllowedChars(rnd.nextInt(numberOfAllowedChars))
    val value = new String(helperArray.map(randomChar))
    TraceId(value)
  }

}

File: src/main/java/com/avast/kluzo/javaapi/TraceId.java
package com.avast.kluzo.javaapi;

import com.avast.kluzo.TraceId$;

import java.util.Objects;

public final class TraceId {

    private final String value;

    public TraceId(String value) {
        this.value = value;
    }

    public static TraceId generate() {
        return new TraceId(TraceId$.MODULE$.generate());
    }

    public String getValue() {
        return value;
    }

    @Override
    public String toString() {
        return String.format("TraceId(%s)", value);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        TraceId traceId = (TraceId) o;
        return Objects.equals(value, traceId.value);
    }

    @Override
    public int hashCode() {
        return Objects.hash(value);
    }

}
